apps/admin/backend/src/tabulation/full_results.ts

```diff
@@ -57,6 +59,33 @@ export function tabulateCastVoteRecords({
   });
 }
 
+/**
+ * Memoized version of `tabulateCastVoteRecords`.
+ */
+const tabulateCastVoteRecordsMemoized = memoize(
+  tabulateCastVoteRecords,
+  ({ electionId, filter, groupBy }) =>
+    hash(
+      {
+        electionId,
+        filter,
+        groupBy,
+      },
+      {
+        unorderedArrays: true,
+        unorderedObjects: true,
+      }
+    )
+);
+
+/**
+ * Clears the memoized results of tabulating cast vote records.
+ */
+export function clearTabulationCache(): void {
```

Comment:
Manual cache invalidation worries me a bit. It's easy to miss a spot, especially as code changes over time (e.g. someone adds a feature that impacts tabulation but forgets to add a call to clear the cache). How big would the impact be if we miss a spot?

A potential way to mitigate this risk would be to tie invalidation to the data itself. E.g. if the data used for tabulation had a `last_updated_at` field that was updated any time the data was changed, you could query that field before loading the full data set and use it to invalidate outdated cache records.

I also want to think through memory management since there are no limits on cache size that I can see. How large is the data being cached? How large can the cache grow before being cleared?

Reply:
The impact would be big (well, equivalent to the same sorts of issues we've had with react query caching) so it's a good call out.

I am also concerned that we could add a function, like deleting individual cast vote records, and not do validation. Two tables influence scanned CVR validation - `cvrs` and `vote_adjudications`. Or actually, we could just use `cvr_files` and `vote_adjudications`. We could have timestamps on rows in both those tables and scan them to clear the cache. Or we could add triggers that update a separately stored, singular timestamp whenever those tables are modified. I think that would be an improvement over the current invalidation.

I don't think we should worry about memory management here. The data being stored is just a JSON representation of the results of an election. There will be one for each report viewed. The amount of data will be less than the amount currently cached on the frontend. cc @eventualbuddha. I think the memory cost of caching here is low and it is worth caching more than one value.

Reply:
I thought about attaching timestamps to the existing data model, but I didn't think that would work as well. For example, say we introduce a feature to delete individual CVR files. You have three CVR files:

- CVR File 1, August 1st
- CVR File 2, August 2nd
- CVR File 3, August 3rd

Let's say you delete the middle file, then you have:

- CVR File 1, August 1st
- CVR File 3, August 3rd

We need some timestamp generated on deletion, which doesn't suit row-by-row timestamps as well.

Reply:
Yeah, I think the solution you came up with makes sense.

I don't have a strong opinion about memory management. I'm sympathetic to @eventualbuddha's point that keeping all the past results cached is unnecessary and also to your point that it's a small amount of memory. Maybe a simple drop-in memoize that only keeps one value would be an easy solution https://github.com/alexreardon/memoize-one

Reply:
The current implementation doesn't clear the cache at all, even if it becomes invalid. That's a change I will make before merging, because there's no reason to keep that data around.

But I don't want to move to a `memoize-one` model because it would get rid of some of the benefit. That would mean that if you print the full election report, go preview another report, and come back and print the full election report again, you'll expend the tabulation cost both times you print, whether that's half a second or half a minute. I'd like to avoid that, and keep more than one cached at a time.

Reply:
Good call!

Reply:
I understand why you want to keep multiple cache values, but the problem as I see it is that the cache never expires. Imagine I call this function thousands of times with different keys, now we're storing thousands of results in memory and never discarding them until the program exits. There should be something that clears the data after a time, no?

Reply:
> That's a change I will make before merging, because there's no reason to keep that data around.

Ah, missed this. SGTM as long as we're not building up an unbounded cache in memory.

Reply:
Well, it would still be unbounded as long as the CVR data doesn't change. But in order to hit the situation you're talking about - with thousands of results - the user would have to generate thousands of custom reports within a power cycle. I don't think that's going to happen.

Reply:
It would be a slow leak, but it's still a leak. And it doesn't have to be reports for a single election. If I never reboot my VxAdmin, I'll slowly leak memory as I generate reports over months and years in the extreme case.

Is this likely? No. But IMO we should not create opportunities for unbounded memory leaks in our application. The easiest solution might be to provide `memoize` with an LRU cache with some fixed maximum count.

Reply:
Provided `memoize` with an `LRUMap`: e7aea2740